src/lexer.c

```diff
@@ -118,6 +82,45 @@ static inline bool lisxdigit(int c) { return testprop(c, MASK(XDIGITBIT)); }
 */
 static inline int ltolower(int c) { return ((c) | ('A' ^ 'a')); }
 
+/*
+Creates a new string object. string objects are interned in a hash set.
+If the string matches a keyword then the reserved attribute will be set to the token id associated
+with the keyword else this attribute will be -1. The hash value of the string is stored the 'hash'
+attribute. Note that we need to allow strings that have embedded 0 character hence the length
+is explicit. But all tokens and reserved keywords are expected to be standard C strings.
+*/
+const struct string_object *raviX_create_string(struct compiler_state *container, const char *input, uint32_t len)
+{
+	struct string_object temp = {.len = len, .str = input, .hash = fnv1_hash_data(input, len), .reserved = -1};
+	struct set_entry *entry = set_search_pre_hashed(container->strings, temp.hash, &temp);
+	if (entry != NULL)
+		/* found the string */
+		return (const struct string_object *)entry->key;
+	else {
+		struct string_object *new_string = raviX_allocator_allocate(&container->string_object_allocator, 0);
+		char *s = raviX_allocator_allocate(&container->string_allocator, len + 1); /* allow for 0 terminator */
+		memcpy(s, input, len);
+		s[len] = 0; /* 0 terminate string, however string may contain embedded 0 characters */
+		new_string->str = s;
+		new_string->len = len;
+		new_string->hash = temp.hash;
+		new_string->reserved = -1;
+		/* Check if this is a keyword, linear search is okay as we do this only when we first
+		 * encounter a keyword
+		 */
+		for (int i = 0; i < ARRAY_SIZE(luaX_tokens); i++) {
+			if (lislalpha(luaX_tokens[i][0]) || luaX_tokens[i][0] == '@') {
```

Comment:
I think this is still incorrect. Our goal is to detect we went past `<eof>`.

Reply:
So the better way is to loop from 0 to (TOK_EOS-FIRST_RESERVED) ?

Reply:
Y I think so

Reply:
It is my mistake - the tokens should be reordered so that we can stop when we hit the "//" token as it makes no sense to treat those terminals as reserved. I checked in Lua code that it works that way. Because I had the new keywords added at the wrong place then this has messed it up. Would you rather I fix this in my code before merging your pull request? Or I ca merge your PR and then fix it myself.

Reply:
I am being stupid here - your version makes it unnecessary to reorder as it skips over the non-terminals anyway.